src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1021,6 +1022,18 @@ public async ValueTask<HttpResponseMessage> SendWithVersionDetectionAndRetryAsyn
                             Debug.Assert(connection is not null || !_http2Enabled);
                             if (connection is not null)
                             {
+                                if (request.IsWebSocketH2Request())
+                                {
+                                    int settingsTimeoutInMilliseconds = 5 * 60 * 1000;
+                                    CancellationTokenSource cts = new CancellationTokenSource(settingsTimeoutInMilliseconds);
+                                    if (await connection.InitialSettingsReceived.WaitWithCancellationAsync(cts.Token).ConfigureAwait(false) && !connection.IsConnectEnabled)
```

Comment:
`cancellationToken` is the request token, right? That might be sufficient. Otherwise us an existing configurable value like the Connect timeout (and don't forget to dispose the CTS, a using statement would help there).
```suggestion
                                    if (await connection.InitialSettingsReceived.WaitWithCancellationAsync(cancellationToken).ConfigureAwait(false) && !connection.IsConnectEnabled)
```